### PR TITLE
Added model for detection and processing

### LINKED ISSUE #54: Add initial logo detection model
Structure:
- in the `models` directory we should create a new `logo_detection` (or similar) directory.
- the `logo_detection` directory should have an `__init__.py` file that:
    - imports everything that should be **exported** from the `logo_detection` directory
    - So if you had `detect.py` that had a `Detector` class and a `process.py` that had a `process` method, the `__init__.py` would look like:
```python
from .detect import Detector
from .process import process
```
- Ideally only 1 class / grouping of functions per file just to keep it cleaner

Requirements:
- Have a look at [the repository](https://huggingface.co/spaces/nathanjc/Logo_detection_YoloV7) and you will see there are a large number of options and different configurations
- For the first iteration we want to pare this back fully (so remove any functionality that is not strictly necessary)
- Given a **path** to a file that should contain an image, the code should run inference on the image and find any bounding boxes of any logo(s) in the image
- The code should not:
    - Draw the bounding boxes to any of the images
    - Write the bounding boxes to any file
    - After running the code, no new files should be made (so for example, `trace` should be turned off)
- The output should be a (possibly empty) list of:
    - coordinates of a bounding box of the logo
        - tbc on exact format. The repo gets them in pixel format and then transforms them to the midpoints of the x and y values and then the width and height values, all as percentages (/100) of the overall width or height respectively
        - we want to follow a similar format where we go on a scalar rather than exact pixel count so that we can allow for dynamic sizing in the front end
    - A confidence score of the accuracy of the logo
    - I would recommend a simple class to hold this data:
```python
class BoundingBox:
    x1: float
    x2: float
    y1: float
    y2: float

class LogoDetectionInference:
    bounding_box: BoundingBox
    confidence: float
```

To consider:
- There may be multiple or no logos per image, so this needs to be accounted for in the processing and response type
- Think carefully about any code you're copying from the repo about whether it is actually needed or is for a feature we aren't supporting etc.
- If we can ideally break the `detect` function into smaller functions that would be great (these are fine to stay in the same file)

### PR DESCRIPTION
**Pull request by liuphui**

### CODE REVIEW

File: ml/models/logo_detection/data.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class BoundingBox:
+    def __init__(self, x: float, y: float, w: float, h: float):
+        self.x = x
+        self.y = y
+        self.w = w
+        self.h = h
+
+class LogoDetectionInference:
+    def __init__(self, boundingBox: BoundingBox, confidence: float):
+        self.boundingBox = boundingBox
```
Comment (sktylr) [inline ml/models/logo_detection/data.py]: Python styling suggests the use of `snake_case` for variables instead of `camelCase`
```suggestion
class LogoDetectionInference:
    def __init__(self, bounding_box: BoundingBox, confidence: float):
        self.bounding_box = bounding_box
```

File: ml/models/logo_detection/detect.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import time
+from pathlib import Path
+import torch
+
+from yolov7_package import Yolov7Detector
+from yolov7_package.models.experimental import attempt_load
+from yolov7_package.utils.general import check_img_size, check_requirements, check_imshow, non_max_suppression, apply_classifier, \
+    scale_coords, xyxy2xywh, strip_optimizer, set_logging, increment_path
+from yolov7_package.utils.torch_utils import select_device, load_classifier, time_synchronized, TracedModel
+from yolov7_package.utils.datasets import LoadStreams, LoadImages
```
Comment (sktylr) [inline ml/models/logo_detection/detect.py]: There are a couple of unused imports here we can remove

File: ml/models/logo_detection/detect.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+from data import BoundingBox, LogoDetectionInference
```
Comment (sktylr) [inline ml/models/logo_detection/detect.py]: To integrate with the rest of the app we will need to use relative imports here to avoid import errors from other files
```suggestion
from .data import BoundingBox, LogoDetectionInference
```

File: ml/models/logo_detection/detect.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+os.system("pip install yolov7-package==0.0.12")
+
```
Comment (sktylr) [inline ml/models/logo_detection/detect.py]: We can remove this since it is included in the update requirements.txt. You may need to merge main into this branch to get the latest requirements.txt changes

File: ml/models/logo_detection/detect.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        # Get stride and imgsz
+        stride, imgsz = self.get_stride_and_imgsz(model, imgsz)
+
+        # Get dataset
+        dataset = self.get_dataset(source, imgsz, stride)
+
+        # Get names
+        names = self.get_names(model)
+
+        detectionInf = self.run_inference(model, device, imgsz, names, half, dataset)
```
Comment (sktylr) [inline ml/models/logo_detection/detect.py]:
```suggestion
        detection_inference = self.run_inference(model, device, imgsz, names, half, dataset)
```
snake case

File: ml/models/logo_detection/detect.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+                        # Store data into the classes
+                        bbox_coordinates = BoundingBox(x=xywh[0], y=xywh[1], w=xywh[2], h=xywh[3])
+                        inference = LogoDetectionInference(boundingBox=bbox_coordinates, confidence=conf.item())
+                        inferences.append(inference)
+
+                # Print time (inference + NMS)
+                print(f'{s}Done. ({(1E3 * (t2 - t1)):.1f}ms) Inference, ({(1E3 * (t3 - t2)):.1f}ms) NMS')
+
+            print(f'Done. ({time.time() - t0:.3f}s)')
```
Comment (sktylr) [inline ml/models/logo_detection/detect.py]: Small thing, but can we change these `prints` to `logger.info` statements. At the top of the file you can put

```python
import logging
... rest of imports

logging.basicConfig()
logger = logging.getLogger()
logger.setLevel(logging.INFO)
```
and then use `logger.info` as you would `print`.

File: ml/models/logo_detection/main.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+from process import process
```
Comment (sktylr) [inline ml/models/logo_detection/main.py]: Same as import above
```suggestion
from .process import process
```

File: ml/models/logo_detection/process.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import torch
+from detect import Detector  # Importing the Detector class from detect.py
```
Comment (sktylr) [inline ml/models/logo_detection/process.py]: Same as import above
```suggestion
from .detect import Detector
```

File: ml/models/logo_detection/process.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+def process(input_image_path):
+
+    source = input_image_path
+    detector = Detector(weights='logo_detection.pt', source=source)
```
Comment (sktylr) [inline ml/models/logo_detection/process.py]: We will need this to be the path to the detection model in the docker build
```suggestion
    detector = Detector(weights='/src/ml/models/logo_detection/logo_detection.pt', source=source)
```
For local testing in isolation to the rest of the API we can update the `main.py` file to use the CWD to find the weights file

**Review by sktylr — CHANGES_REQUESTED** [review]
Nice work Peng! Just a few small tweaks but is all working for me locally.

To get the CI to pass, you will need to run `make format` in the root directory

**Comment by sktylr** [comment]
We can also remove the `.gitkeep` file in the `ml/models/` directory, it's used to maintain repository structure before files are added

File: ml/models/logo_detection/__init__.py [hunk 10]
```diff
@@ -0,0 +1,3 @@
+# from .data import BoundingBox, LogoDetectionInference
+# from .detect import Detector
+# from .process import process
```
Comment (sktylr) [inline ml/models/logo_detection/__init__.py]: You can actually uncomment these - when you merge main into this branch these linting errors will go away

**Review by sktylr — APPROVED** [review]
Great work 🚀
